### ISSUE TITLE
The twisted.web server should not ignore "Expect: 100-continue"

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4673]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#4673|
|Type|defect|
|Created|2010-10-05 19:08:16Z|
|Branch|https://github.com/twisted/twisted/tree/100-continue-4673|

From the RFC: "The purpose of the 100 (Continue) status ... is to allow a client that is sending a request message with a request body to determine if the origin server is willing to accept the request (based on the request headers) before the client sends the request body."

See http://www.w3.org/Protocols/rfc2616/rfc2616-sec8.html#sec8.2.3 for details.

Supporting the "Expect: 100-continue" header on an API level is currently difficult, since twisted.web always reads the full request body before handing the request to business logic code. Until we can allow custom Resources to handle requests before the body is read, we should at the very least return the appropriate "100 (Continue)" response to the HTTP client so it can continue.

<details><summary>Searchable metadata</summary>

```
trac-id__4673 4673
type__defect defect
reporter__itamar itamar
priority__low low
milestone__ 
branch__branches_100_continue_4673 branches/100-continue-4673
branch_author__itamarst itamarst
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1286305696000000 1286305696000000
changetime__1316894649000000 1316894649000000
version__None None
owner__itamar itamar
cc__msteder
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst set owner to @jyknight|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_4"></a>|@itamarst commented|
|-|-|

(In [32482]) Branching to '100-continue-4673'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst commented|
|-|-|

This is an implementation of the simple, just respond with 100 continue, suggestion. A user-exposed API depends on [#4688](https://github.com/twisted/twisted/issues/4688). I tested with curl, and without this POSTs take an extra 1 second as curl waits for the 100 continue, decides it's not going to happen and then switches to normal processing.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_6"></a>|@itamarst commented|
|-|-|

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/100-continue-4673 is the buildbot run, just started, hopefully they all pass :).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @itamarst|
|-|-|

  1. I think that writing out a 100-Continue response as soon as the "Expect: 100-Continue" request header is received has a minor problem.  A later request header may be unparseable and result in a 400 response; for example, a multipart/form-data request without a content-disposition field results in a 400 response.  Another case which may result in a 400 response is the case where the maximum number of request headers is exceeded.  It's not the end of the world if we send a 100 and then a 400, probably, but it would probably be better to avoid doing so.
  1. It's not immediately obvious that this works correctly with pipelined requests.  I think it does, but it would be nice to see a unit test demonstrating that the response to an earlier request isn't corrupted by a later request which includes an "Expect: 100-Continue".
  1. `test_HTTP10` docstring typos _Except_ for _Expect_

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_8"></a>|@itamarst removed owner|
|-|-|

All 3 comments were addressed. As a result of the change, the Expect header will now (as in trunk) be exposed to user code; this can be changed if necessary.

Test run, in progress: http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/100-continue-4673

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @itamarst|
|-|-|

Thanks, I like the look of this.  One thing though, for the behavior exercised by the new `test_expect100ContinueWithPipelining`, I was thinking of the case where requests arrive in the opposite order as they do in the current implementation, and where the first request has an asynchronous response which lasts past the time when the second request is received.  If you feel like adding that one, cool, otherwise I don't know if the new one that's there now offers much - I suppose it's a another case but it doesn't seem like one that's very likely to different.

Also, the news fragment says _Twisted.web's server code_ but it should probably be _twisted.web's server code_ or even _twisted.web.server_.

Address to your satisfaction and merge.  Thanks.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_10"></a>|@itamarst commented|
|-|-|

I'm just going to merge as is (with updated news file).

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_11"></a>|@itamarst set status to `closed`|
|-|-|

(In [32596]) Merge 100-continue-4673.

Fixes: [#4673](https://github.com/twisted/twisted/issues/4673)
Author: itamarst
Review: exarkun

twisted.web server code now responds to 100-continue headers, although this is not controllable by user code yet.
